### DISCUSSION TITLE
Differences in performance between DuckDB native and Parquet

### DISCUSSION THREAD

**Discussion by cpcloud** [duckdb/duckdb#10161] [General]
### What happens?

I wrote a blog post comparing DuckDB to Pandas and Dask [here](https://ibis-project.org/posts/pydata-performance/), and I thought it would be interesting to compare DuckDB's native file format to Parquet just for my own curiosity.

I would expect DuckDB's native file format to be as least as fast as Parquet (without knowing a great deal about the native format), but what I see is about 2-3x worse with the native format than with Parquet.

The workload finishes in about 30-40s with Parquet and anywhere from 1m to 1m30s with the native format. Memory use is also much higher with the native format: around 60GB for native and around 3 GB for Parquet.

### To Reproduce

The blog shows the machine I am running on, but I believe this would be reproducible on any machine, the main challenge is that you need about ~90GB of RAM to create the DuckDB file from the Parquet files.

### 1. Download data

In any event, the first step is to download one of the datasets in the release here: https://github.com/pypi-data/data/releases.

I did this using `gh`:

```bash
$ gh release download -R pypi-data/data 2023-12-09-03-06
```

By the time someone reads this the data may be gone, but you can list release using

```bash
$ gh release list -R pypi-data/data
```

and grab any release's data.

### 2. Create the DuckDB database file

Run this command from the directory to which you download the Parquet files:

```bash
$ duckdb pypi.ddb <<< "CREATE TABLE pypi AS FROM read_parquet('*.parquet')"
```

This command takes a while, the data are about 26 GiB on disk and the DucKDB file ends up being about 4x as big.

I don't particularly care about the increase in disk usage as long as I can get better performance than with Parquet files.

### 3. Run the SQL queries

Here's the ibis-generated query for Parquet.

<details>

<summary>Parquet</summary>

```sql
WITH t0 AS (
  SELECT
    DATE_TRUNC('month', t2.uploaded_on) AS month,
    NULLIF(
      REPLACE(
        REPLACE(
          REPLACE(
            REGEXP_REPLACE(
              REGEXP_REPLACE(
                REGEXP_EXTRACT(t2.path, '\.([a-z0-9]+)$', CAST(1 AS INT)),
                'cxx|cpp|cc|c|hpp|h',
                'C/C++',
                'g'
              ),
              '^f.*$',
              'Fortran',
              'g'
            ),
            'rs',
            'Rust'
          ),
          'go',
          'Go'
        ),
        'asm',
        'Assembly'
      ),
      ''
    ) AS ext,
    COUNT(DISTINCT t2.project_name) AS project_count
  FROM read_parquet('*.parquet') AS t2
  WHERE
    REGEXP_MATCHES(t2.path, '\.(asm|c|cc|cpp|cxx|h|hpp|rs|[Ff][0-9]{0,2}(?:or)?|go)$')
    AND NOT REGEXP_MATCHES(t2.path, '(^|/)test(|s|ing)')
    AND NOT CONTAINS(t2.path, '/site-packages/')
  GROUP BY
    1,
    2
)
SELECT
  t1.month,
  t1.ext,
  t1.project_count
FROM (
  SELECT
    t0.month AS month,
    t0.ext AS ext,
    t0.project_count AS project_count
  FROM t0
  WHERE
    NOT t0.ext IS NULL
) AS t1
ORDER BY
  t1.month DESC,
  t1.project_count DESC
```

</details>

And the same one but with `main.pypi` after creating the `pypi.ddb` file:

<details>

<summary>DuckDB native</summary>

```sql
WITH t0 AS (
  SELECT
    DATE_TRUNC('month', t2.uploaded_on) AS month,
    NULLIF(
      REPLACE(
        REPLACE(
          REPLACE(
            REGEXP_REPLACE(
              REGEXP_REPLACE(
                REGEXP_EXTRACT(t2.path, '\.([a-z0-9]+)$', CAST(1 AS INT)),
                'cxx|cpp|cc|c|hpp|h',
                'C/C++',
                'g'
              ),
              '^f.*$',
              'Fortran',
              'g'
            ),
            'rs',
            'Rust'
          ),
          'go',
          'Go'
        ),
        'asm',
        'Assembly'
      ),
      ''
    ) AS ext,
    COUNT(DISTINCT t2.project_name) AS project_count
  FROM main.pypi AS t2
  WHERE
    REGEXP_MATCHES(t2.path, '\.(asm|c|cc|cpp|cxx|h|hpp|rs|[Ff][0-9]{0,2}(?:or)?|go)$')
    AND NOT REGEXP_MATCHES(t2.path, '(^|/)test(|s|ing)')
    AND NOT CONTAINS(t2.path, '/site-packages/')
  GROUP BY
    1,
    2
)
SELECT
  t1.month,
  t1.ext,
  t1.project_count
FROM (
  SELECT
    t0.month AS month,
    t0.ext AS ext,
    t0.project_count AS project_count
  FROM t0
  WHERE
    NOT t0.ext IS NULL
) AS t1
ORDER BY
  t1.month DESC,
  t1.project_count DESC
```

</details>

### OS:

Linux (NixOS) x86_64

### DuckDB Version:

0.9.2

### DuckDB Client:

Python

### Full Name:

Phillip Cloud

### Affiliation:

Voltron Data

### Have you tried this on the latest `main` branch?

I have tested with a release build (and could not test with a main build)

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] Yes, I have

**Comment by cpcloud** [comment]
Here's the [`btop`](https://github.com/aristocratos/btop) timeline with DuckDB native:

![image](https://github.com/duckdb/duckdb/assets/417981/b967a43d-c037-494e-9d9e-e69c63187505)

The same with Parquet:

![image](https://github.com/duckdb/duckdb/assets/417981/31a13492-27eb-4259-985e-9f93828b7bee)

**Comment by cpcloud** [comment]
Just a note, I updated the SQL queries due to an extra slash in in the regular expression patterns. These are escaped in Python which is where the extra slash is coming from.

I've verified that the Python code and the SQL queries above (with extra slashes removed) produce the same results:

```
❯ ./result/bin/duckdb /data/pypi-duckdb/pypi.ddb < /data/test.sql
100% ▕████████████████████████████████████████████████████████████▏
┌────────────┬──────────┬───────────────┐
│   month    │   ext    │ project_count │
│    date    │ varchar  │     int64     │
├────────────┼──────────┼───────────────┤
│ 2023-11-01 │ C/C++    │           836 │
│ 2023-11-01 │ Rust     │           190 │
│ 2023-11-01 │ Fortran  │            48 │
│ 2023-11-01 │ Go       │            33 │
│ 2023-11-01 │ Assembly │            10 │
│ 2023-10-01 │ C/C++    │          1331 │
│ 2023-10-01 │ Rust     │           250 │
│ 2023-10-01 │ Fortran  │            95 │
│ 2023-10-01 │ Go       │            46 │
│ 2023-10-01 │ Assembly │            23 │
│ 2023-09-01 │ C/C++    │          1456 │
│ 2023-09-01 │ Rust     │           239 │
│ 2023-09-01 │ Fortran  │            60 │
│ 2023-09-01 │ Go       │            42 │
│ 2023-09-01 │ Assembly │            18 │
│ 2023-08-01 │ C/C++    │          1088 │
│ 2023-08-01 │ Rust     │           237 │
│ 2023-08-01 │ Fortran  │            57 │
│ 2023-08-01 │ Go       │            34 │
│ 2023-08-01 │ Assembly │            15 │
│     ·      │    ·     │             · │
│     ·      │    ·     │             · │
│     ·      │    ·     │             · │
│ 2006-10-01 │ Fortran  │             1 │
│ 2006-09-01 │ C/C++    │            12 │
│ 2006-08-01 │ C/C++    │             8 │
│ 2006-08-01 │ Fortran  │             1 │
│ 2006-07-01 │ C/C++    │            10 │
│ 2006-06-01 │ C/C++    │             5 │
│ 2006-05-01 │ C/C++    │             4 │
│ 2006-04-01 │ C/C++    │             4 │
│ 2006-03-01 │ C/C++    │             8 │
│ 2006-02-01 │ C/C++    │             3 │
│ 2006-01-01 │ C/C++    │             5 │
│ 2005-12-01 │ C/C++    │             5 │
│ 2005-11-01 │ C/C++    │             6 │
│ 2005-10-01 │ C/C++    │             2 │
│ 2005-09-01 │ C/C++    │             2 │
│ 2005-08-01 │ C/C++    │             7 │
│ 2005-07-01 │ C/C++    │             4 │
│ 2005-05-01 │ C/C++    │             1 │
│ 2005-04-01 │ C/C++    │             1 │
│ 2005-03-01 │ C/C++    │             1 │
├────────────┴──────────┴───────────────┤
│ 799 rows (40 shown)         3 columns │
└───────────────────────────────────────┘
```

☝🏻 This run is from `main` as of a few minutes ago cde7dca65fa847d794827252e5dd655b92db4db5 and both the runtime and memory issues still exist there.

**Comment by hannes** [comment]
Thanks for the investigations, some performance differences are unavoidable here due to the differences in format and reader implementation, especially for a complex query like that. It would be easier to investigate if the query were much simpler.

**Reply by cpcloud** [reply]
Makes sense! Perhaps the language on the website is promising a bit too much? (https://duckdb.org/docs/guides/performance/file-formats#reasons-against-querying-parquet-files)

![image](https://github.com/duckdb/duckdb/assets/417981/0765450e-8e56-4cc2-9f37-208c8c790112)

**Reply by alexey-lv** [reply]
So in general what should be the expectations when querying Parquet files v.s. native format? Any guidelines / recommendations if the aggregation query performance is important?

**Comment by sabman** [comment]
A bit late to this - I see no mention of any indicies in the discussions or the blog posts. Sorry if I missed this, but were indicies created during this comparison and if so where? If they were not created then, is this really the best way to assess the performance advantages associated with a duckdb native format database?

**Comment by jgarveyanalytics** [comment]
This is an interesting 'thread', no pun intended. Have been flip flopping on this topic as well. Initially had a parquet dataset partitioned by year/month/day/location/*.parquet.

The only instances where using hive-structured parquet was faster were:

- Reading a single partition, i.e. year=2025/month=12/day=10/location=US.
- Writing. Using PARTITION BY when saving data to .parquet is lightning fast.

In all other cases, using a .duckdb file wiped the floor with .parquet. Sometimes up to 100x.

I do appreciate my hive-style folder structure is over-partitioned. Opening and closing hundreds of thousands of .parquet files from my experience is painfully slow for obvious reasons. I think year/month/location would have been sufficient. Using a HP Z840 with 24 physical cores and 512GB SATA SSD, for the record.

Would love to hear others thoughts. Can indices get the same purchase for filtering as parquet partition pruning? What is the equivalent of Oracle / SQL server partitioning in duckdb (.duckdb, not parquet).
